packages/flutter/lib/src/material/material_state_mixin.dart

```diff
@@ -0,0 +1,142 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'material_state.dart';
+
+/// Mixin for [State] classes that require knowledge of changing [MaterialState]
+/// values for their child widgets.
+///
+/// This mixin does nothing by mere application it to a [State] class, but is
+/// helpful when writing `build` methods that include child [InkWell],
+/// [GestureDetector], [MouseRegion], or [Focus] widgets. Instead of manually
+/// creating handlers for each type of user interaction, such [State] classes can
+/// instead provide a `ValueChanged<bool>` function and allow [MaterialStateMixin]
+/// to manage the set of active [MaterialState]s, and the calling of [setState]
+/// as necessary.
+///
+/// {@tool snippet}
+/// ```dart
+/// class MyWidget extends StatefulWidget {
+///   const MyWidget(this.color, {required this.child, Key? key}) : super(key: key);
+///   final MaterialStateColor color;
+///   final Widget child;
+///   @override
+///   State createState() => MyWidgetState();
+/// }
+///
+/// class MyWidgetState extends State<MyWidget> with MaterialStateMixin<MyWidget> {
+///   @override
+///   Widget build(BuildContext context) {
+///     return InkWell(
+///       onFocusChange: updateMaterialState(MaterialState.focused),
+///       child: Container(
+///         color: widget.color.resolve(materialStates),
+///         child: widget.child,
+///       ),
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+@optionalTypeArgs
+mixin MaterialStateMixin<T extends StatefulWidget> on State<T> {
+  /// Managed set of active [MaterialState] values; designed to be passed to
+  /// [MaterialStateProperty.resolve] methods.
+  ///
+  /// To mutate and have [setState] called automatically for you, use
+  /// [setMaterialState], [addMaterialState], or [removeMaterialState]. Directly
+  /// mutating the set is possible, and may be necessary if you need to alter its
+  /// list without calling [setState] (and thus triggering a re-render).
+  ///
+  /// To check for a single condition, convenience getters [isPressed], [isHovered],
+  /// [isFocused], etc, are available for each [MaterialState] value.
+  Set<MaterialState> materialStates = <MaterialState>{};
+
+  /// Callback factory which accepts a [MaterialState] value and returns a
+  /// closure to mutate [materialStates] and call [setState].
+  ///
+  /// {@tool snippet}
+  /// Usage:
```

Comment:
Usually we include a little more information her describing what the example showcases.